Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,249 @@
+   9. All of the mechanisms described above for memory management are not only memory providers, but memory users as well, because they rely on metadata that cannot be pre-allocated and must be adjusted accordingly to the memory transactions happening, as the system evolves over time; therefore they also consume memory, generating an overhead, and being exposed themselves to interference.
+   10. The Linux kernel provides means to limit certain requests a process might present; for example with cgroups it is possible to create logical memory "bubbles" that cannot grow beyond a set size, and associate processes to them, that share the collective limit within the bubble. But this does not much toward separating how the kernel might use the underlying memory, besides setting the constraint as described.
+      
+#### **Safety-Oriented consideration**
+The following considerations are of a more deductive nature.
+
+1. Because of the way pages, fractions and multiples of them are allocated, freed, cached, recovered, there is a complex interaction between system components at various layers.
+2. Even using cgroups, it is not possible to segregate interaction at the low level between components with differnet level of safety qualification (e.g. a QM container can and most likely will affect the recirculation of pages related to an ASIL one)
+3. Because of the nature of memory management, it must be expected that memory management mechanisms will  interfere with safe processes, either due to a bug or due to the interference toward the metadata they rely on. For example, the memory management might hand over to a requesting entity a memory page that is currently already in use either by a device driver or by a userspace process playing a role in a safety use case.
+4. Still due to te complex interaction between processes, kernel drivers and other kernel code, it is practically impossible to qualify the kernel as safe through positive testing alone, because it is impossible to validate all the possible combinations, and it is equally impossible to assess the overall test coverage and the risk associated with not reaching 100%. The only reliable way to test is to use negative testing (simulating a certain type of interference) and confirming that the system response is consistent with expectations (e.g detect the interference, in case of ASILB requirements). And even then, the only credible claim that can be made is that, given the simulated type of interference, on the typology of target employed, the reaction is aligned with the requirements. Other types of targets will require further ad-hoc negative testing.
+5. Linux Kernel mechanisms like SELinux and cgroups/containers do not offer any protection against interference originating from the kernel itself.
+6. The Linux Kernel must be assumed to be QM, unless specific subsystems are qualified through both positive and negative testing.
```

Comment:
Above I would say "Kernel code must be assumed NSR (non safety related), unless the integrator performs additional verification activities to claim a specific systematic capability target (including QM)"

Do we have upstream evidences of a QMS claim ?

Reply:
I have been overenthusiastic :-D 
The problem I see with the proposed change is that it waters down the statement excessively.
I can reword it as "... must be assumed to not be safe; possibly QM at best ..."
But "veriication" is too bland because it becomes vvery subjective, while the point here is to stress he need for both positive and negative testing.

Reply:
I made the change, so closing this as obsolete